### DISCUSSION TITLE
Thread Context in asynchronous programs

### DISCUSSION THREAD

**Discussion by jfontsaballs** [apache/logging-log4j2#1826] [Q&A]
I'm writing a program that makes heavy use of Kotlin coroutines (it is a single threaded application handling a few tens of concurrent jobs) and uses Log4j2. 

If I'm not mistaken, information pushed into Thread Context is stored in ThreadLocal storage. Thus, I believe I can't use Thread Context in my application. Is it right?

Is there any way that I can provide this Context (map and/or stack) information to the logging call, without going through ThreadLocal storage?

**Comment by jvz** [comment]
This might be tricky to figure out, but I think it's possible in theory at least. In Log4j API, there are the `ThreadContextMap` and `ThreadContextStack` interfaces which can be overridden to use a different implementation. However, the coroutine context provided in the Log4j Kotlin API handles propagation of these thread locals for you. A non-thread-local version of the same would still require a `CoroutineContext` I believe, but it would require custom TCM/TCS classes like I said.

Edit: see https://github.com/apache/logging-log4j-kotlin/blob/main/log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/CoroutineThreadContext.kt for more info.

**Reply by jfontsaballs** [reply]
I'll take a look into it, thank you.
